File: android/connector/Connector.java
package com.dodotdo.base.connector;

import com.dodotdo.base.util.USON;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class Connector {

    protected final Map<String, List<Method>> methodMap = new HashMap<>();
    protected int sendKey = 0;

    protected void runMethods(String methodName, Object data) {
        List<Method> methods = methodMap.get(methodName);
        if (methods != null) {
            try {
                for (Method method : methods) {
                    method.handle(data instanceof JSONObject ? USON.unpack((JSONObject) data) : data, (Object callbackData) -> {
                    });
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }

    protected void runMethods(String methodName, Object data, int sendKey) {
        List<Method> methods = methodMap.get(methodName);
        if (methods != null) {
            try {
                for (Method method : methods) {
                    method.handle(data instanceof JSONObject ? USON.unpack((JSONObject) data) : data, (Object callbackData) -> {
                        send("__CALLBACK_" + sendKey, callbackData);
                    });
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }

    public void on(String methodName, Method method) {
        List<Method> methods = this.methodMap.get(methodName);
        if (methods == null) {
            methods = new ArrayList<>();
        }
        methods.add(method);
        this.methodMap.put(methodName, methods);
    }

    public void off(String methodName, Method method) {
        List<Method> methods = this.methodMap.get(methodName);
        if (methods != null) {
            methods.remove(method);
        }
    }

    public void off(String methodName) {
        this.methodMap.remove(methodName);
    }

    public abstract void send(String methodName);

    public abstract void send(String methodName, Object data);

    public abstract void send(String methodName, Callback callback);

    public abstract void send(String methodName, Object data, Callback callback);
}

File: ios/Connector/Connector.swift
import Foundation
import SwiftyJSON

protocol Connector: class {
  
  var methodMap: [String : [Method]] { get set }
  
  func send(methodName: String)
  func send(methodName: String, data: JSON?)
  func send(methodName: String, callback: Callback)
  func send(methodName: String, data: JSON?, callback: Callback)
}

extension Connector {
  
  func runMethods(methodName: String, data: JSON?) {
    let methods = methodMap[methodName]
    if methods != nil {
      for method in methods! {
        method.exec(data, Callback { (callbackData: JSON?) -> Void in })
      }
    }
  }
  
  func runMethods(methodName: String, data: JSON?, sendKey: UInt) {
    let methods = methodMap[methodName]
    if methods != nil {
      for method in methods! {
        method.exec(data, Callback { (callbackData: JSON?) -> Void in
          self.send(methodName: "__CALLBACK_\(sendKey)", data: callbackData)
        })
      }
    }
  }
  
  func on(_ methodName: String, _ method: Method) {
    var methods = methodMap[methodName]
    if methods != nil {
      methods = []
    }
    methods!.append(method)
    methodMap[methodName] = methods
  }
  
  func off(_ methodName: String, _ method: Method) {
    var methods = methodMap[methodName]
    if methods != nil {
      methods!.removeAll(where: {$0 === method})
    }
  }
  
  func off(_ methodName: String) {
    methodMap.removeValue(forKey: methodName);
  }
}
